### PR TITLE
Use helper extension to simplify code.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -52,13 +52,12 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                 return;
             }
 
-            if (!variableDeclarator.IsParentKind(SyntaxKind.VariableDeclaration) ||
+            if (!variableDeclarator.IsParentKind(SyntaxKind.VariableDeclaration, out VariableDeclarationSyntax variableDeclaration) ||
                 !variableDeclarator.Parent.IsParentKind(SyntaxKind.LocalDeclarationStatement))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]:
```suggestion
                !variableDeclaration.IsParentKind(SyntaxKind.LocalDeclarationStatement))
```

File: src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs [hunk 2]
```diff
@@ -287,13 +287,13 @@ bool TryProcessLocalDeclarationStatement(LocalDeclarationStatementSyntax localDe
                         .Where(statement => Equals(semanticModel.GetEnclosingSymbol(
                             statement.SpanStart, cancellationToken), memberDeclarationSymbol)).Count();
 
-                    if (forEachInfo.ForEachStatement.IsParentKind(SyntaxKind.Block) &&
+                    if (forEachInfo.ForEachStatement.IsParentKind(SyntaxKind.Block, out BlockSyntax block) &&
                         forEachInfo.ForEachStatement.Parent.Parent == memberDeclarationSyntax)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs]:
```suggestion
                        block.Parent == memberDeclarationSyntax)
```

File: src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs [hunk 3]
```diff
@@ -94,12 +93,10 @@ protected override bool IsArrayElementType(ExpressionSyntax expression)
         protected override bool IsInInterfaceList(ExpressionSyntax expression)
         {
             if (expression is TypeSyntax &&
-                expression.Parent is BaseTypeSyntax &&
-                expression.Parent.IsParentKind(SyntaxKind.BaseList) &&
-                ((BaseTypeSyntax)expression.Parent).Type == expression)
+                expression.Parent is BaseTypeSyntax baseType &&
+                expression.Parent.IsParentKind(SyntaxKind.BaseList, out BaseListSyntax baseList) &&
```
Comment (sharwell) [inline src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs]:
```suggestion
                baseType.IsParentKind(SyntaxKind.BaseList, out BaseListSyntax baseList) &&
```

File: src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs [hunk 4]
```diff
@@ -114,11 +111,9 @@ protected override bool IsInInterfaceList(ExpressionSyntax expression)
             }
 
             if (expression is TypeSyntax &&
-                expression.IsParentKind(SyntaxKind.TypeConstraint) &&
-                expression.Parent.IsParentKind(SyntaxKind.TypeParameterConstraintClause))
+                expression.IsParentKind(SyntaxKind.TypeConstraint, out TypeConstraintSyntax typeConstraint) &&
+                expression.Parent.IsParentKind(SyntaxKind.TypeParameterConstraintClause, out TypeParameterConstraintClauseSyntax constraintClause))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs]:
```suggestion
                typeConstraint.IsParentKind(SyntaxKind.TypeParameterConstraintClause, out TypeParameterConstraintClauseSyntax constraintClause))
```

File: src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceService.cs [hunk 5]
```diff
@@ -32,9 +32,9 @@ public CSharpImplementInterfaceService()
         {
             if (!cancellationToken.IsCancellationRequested)
             {
-                if (node is TypeSyntax interfaceNode && interfaceNode.Parent is BaseTypeSyntax &&
+                if (node is TypeSyntax interfaceNode && interfaceNode.Parent is BaseTypeSyntax baseType &&
                     interfaceNode.Parent.IsParentKind(SyntaxKind.BaseList) &&
```
Comment (sharwell) [inline src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceService.cs]:
```suggestion
                    baseType.IsParentKind(SyntaxKind.BaseList) &&
```

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 6]
```diff
@@ -327,22 +316,16 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
...
-            out LocalDeclarationStatementSyntax localDeclaration)
+            AnonymousFunctionExpressionSyntax anonymousFunction, out LocalDeclarationStatementSyntax localDeclaration)
         {
             // Type t = null;
             // t = <anonymous function>
-            if (anonymousFunction.IsParentKind(SyntaxKind.SimpleAssignmentExpression) &&
-                anonymousFunction.Parent.IsParentKind(SyntaxKind.ExpressionStatement) &&
-                anonymousFunction.Parent.Parent.IsParentKind(SyntaxKind.Block))
+            if (anonymousFunction.IsParentKind(SyntaxKind.SimpleAssignmentExpression, out AssignmentExpressionSyntax assignment) &&
+                anonymousFunction.Parent.IsParentKind(SyntaxKind.ExpressionStatement, out ExpressionStatementSyntax expressionStatement) &&
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]:
```suggestion
                assignment.IsParentKind(SyntaxKind.ExpressionStatement, out ExpressionStatementSyntax expressionStatement) &&
```

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs [hunk 7]
```diff
@@ -327,22 +316,16 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext, INamedTyp
...
+                anonymousFunction.Parent.Parent.IsParentKind(SyntaxKind.Block, out BlockSyntax block))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]:
```suggestion
                expressionStatement.IsParentKind(SyntaxKind.Block, out BlockSyntax block))
```

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs [hunk 8]
```diff
@@ -482,9 +482,9 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
                 var symbol = semanticModel.GetSymbolInfo(memberAccess.Name).Symbol;
 
                 if (previousToken.Kind() == SyntaxKind.OpenParenToken &&
-                    previousToken.Parent.IsKind(SyntaxKind.ParenthesizedExpression) &&
+                    previousToken.Parent.IsKind(SyntaxKind.ParenthesizedExpression, out ParenthesizedExpressionSyntax parenExpr) &&
                     !previousToken.Parent.IsParentKind(SyntaxKind.ParenthesizedExpression) &&
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs]:
```suggestion
                    !parenExpr.IsParentKind(SyntaxKind.ParenthesizedExpression) &&
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 9]
```diff
@@ -955,12 +955,11 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
             // Might be an incomplete conditional expression or an incomplete declaration of a method returning a nullable type.
             // Bind T to see if it is a type. If it is we don't show signature help.
             if (name.IsParentKind(SyntaxKind.LessThanExpression) &&
-                name.Parent.IsParentKind(SyntaxKind.ConditionalExpression) &&
+                name.Parent.IsParentKind(SyntaxKind.ConditionalExpression, out ConditionalExpressionSyntax conditional) &&
                 name.Parent.Parent.IsParentKind(SyntaxKind.ExpressionStatement) &&
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
                conditional.IsParentKind(SyntaxKind.ExpressionStatement) &&
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 10]
```diff
@@ -955,12 +955,11 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
...
                 name.Parent.Parent.Parent.IsParentKind(SyntaxKind.GlobalStatement))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
                conditional.Parent.IsParentKind(SyntaxKind.GlobalStatement))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 11]
```diff
@@ -1016,10 +1015,9 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
 
             if (token.IsKind(SyntaxKind.CloseBracketToken) &&
                 token.Parent.IsKind(SyntaxKind.AttributeList) &&
-                token.Parent.IsParentKind(SyntaxKind.Parameter) &&
+                token.Parent.IsParentKind(SyntaxKind.Parameter, out ParameterSyntax parameter) &&
                 token.Parent.Parent.Parent.IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParameterList(includeOperators))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
                parameter.Parent.IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParameterList(includeOperators))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 12]
```diff
@@ -1031,7 +1029,7 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
                 token.Parent.IsKind(SyntaxKind.Parameter) &&
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
                token.Parent.IsKind(SyntaxKind.Parameter, out parameter) &&
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 13]
```diff
@@ -1031,7 +1029,7 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
...
                 token.Parent.Parent.IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParameterList(includeOperators))
             {
-                var parameter = (ParameterSyntax)token.Parent;
+                parameter = (ParameterSyntax)token.Parent;
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 14]
```diff
@@ -1031,7 +1029,7 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
                 token.Parent.IsKind(SyntaxKind.Parameter) &&
                 token.Parent.Parent.IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParameterList(includeOperators))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]:
```suggestion
                parameter.Parent.IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParameterList(includeOperators))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs [hunk 15]
```diff
@@ -332,14 +331,16 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
             //    as the interface type itself.
             if (type != null)
             {
-                if (type.Parent is BaseTypeSyntax && type.Parent.IsParentKind(SyntaxKind.BaseList) && ((BaseTypeSyntax)type.Parent).Type == type)
+                if (type.Parent is BaseTypeSyntax baseType &&
+                    type.Parent.IsParentKind(SyntaxKind.BaseList, out BaseListSyntax baseList) &&
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs]:
```suggestion
                    baseType.IsParentKind(SyntaxKind.BaseList, out BaseListSyntax baseList) &&
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs [hunk 16]
```diff
@@ -348,10 +349,9 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
 
             // 4) The type of a constant must be at least as accessible as the constant itself.
             // 5) The type of a field must be at least as accessible as the field itself.
-            if (type.IsParentKind(SyntaxKind.VariableDeclaration) &&
+            if (type.IsParentKind(SyntaxKind.VariableDeclaration, out VariableDeclarationSyntax variableDeclaration) &&
                 type.Parent.IsParentKind(SyntaxKind.FieldDeclaration))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs]:
```suggestion
                variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs [hunk 17]
```diff
@@ -360,10 +360,9 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
             if (type.IsParentKind(SyntaxKind.ObjectCreationExpression) &&
                 type.Parent.IsParentKind(SyntaxKind.EqualsValueClause) &&
                 type.Parent.Parent.IsParentKind(SyntaxKind.VariableDeclarator) &&
-                type.Parent.Parent.Parent.IsParentKind(SyntaxKind.VariableDeclaration) &&
+                type.Parent.Parent.Parent.IsParentKind(SyntaxKind.VariableDeclaration, out variableDeclaration) &&
                 type.Parent.Parent.Parent.Parent.IsParentKind(SyntaxKind.FieldDeclaration))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs]:
```suggestion
                variableDeclaration.IsParentKind(SyntaxKind.FieldDeclaration))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs [hunk 18]
```diff
@@ -420,10 +419,9 @@ public static ISet<INamespaceSymbol> GetUsingNamespacesInScope(this SemanticMode
             }
 
             // 8) The type of an event must be at least as accessible as the event itself.
-            if (type.IsParentKind(SyntaxKind.VariableDeclaration) &&
+            if (type.IsParentKind(SyntaxKind.VariableDeclaration, out variableDeclaration) &&
                 type.Parent.IsParentKind(SyntaxKind.EventFieldDeclaration))
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SemanticModelExtensions.cs]:
```suggestion
                variableDeclaration.IsParentKind(SyntaxKind.EventFieldDeclaration))
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs [hunk 19]
```diff
@@ -1719,10 +1698,9 @@ private IEnumerable<TypeInferenceInfo> InferTypeInNameColon(NameColonSyntax name
                                     ? delegateType.TypeArguments[0]
                                     : this.Compilation.ObjectType;
 
-                                if (IsUnusableType(typeArg) && argumentExpression is LambdaExpressionSyntax)
+                                if (IsUnusableType(typeArg) && argumentExpression is LambdaExpressionSyntax lambdaExpression)
```
Comment (sharwell) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: 📝 This one should have been caught by our existing analyzer. Can you file a bug if it was not?

**Comment by CyrusNajmabadi** [comment]
Will make the changes in a followup PR :)

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]: fixed in next PR.

File: src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs]: Done.

File: src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs]: done.

File: src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/GenerateType/CSharpGenerateTypeService.cs]: done.

File: src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceService.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceService.cs]: done.

File: src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UseLocalFunction/CSharpUseLocalFunctionDiagnosticAnalyzer.cs]: done.

File: src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/CSharp/Portable/Simplification/Simplifiers/ExpressionSimplifier.cs]: done

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: done.

**Comment by CyrusNajmabadi** [comment]
All feedback accounted for.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs]: will check!

**Pull request merged**